xarray/core/dataarray.py

```diff
@@ -115,6 +115,28 @@ def _infer_coords_and_dims(
     return new_coords, dims
 
 
+def _check_data_shape(data, coords, dims):
+    if data.size == 1 and coords is not None:
```

Comment:
This should be `data.shape == ()` to identify scalar arrays, not merely size 1 arrays.

Reply:
You are right, the use of size here was not raising an error when doing something like `xarray.DataArray(np.array([[[1.0]]]), coords=[('x', np.arange(3)), ('y', np.arange(7)])` even though the data shape didn't match the `coords`.